### ISSUE TITLE
build 0.4.8 error 

### ISSUE THREAD

**Issue by 12sunflowers** [freeorion/freeorion#2266]
CentOS 7
**cmake**
```
[root@localhost src-tarball]# cmake .
-- Build type CMAKE_BUILD_TYPE set to Release
-- Use CCache for Unix Makefiles called with: ccache program  
-- Boost version: 1.59.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   iostreams
--   locale
--   log
--   log_setup
--   regex
--   serialization
--   signals
--   system
--   thread
--   chrono
--   atomic
-- Boost version: 1.59.0
-- Found the following Boost libraries:
--   python
CMake Deprecation Warning at GG/CMakeLists.txt:29 (cmake_policy):
  The OLD behavior for policy CMP0044 will be removed from a future version
  of CMake.

  The cmake-policies(7) manual explains that the OLD behaviors of all
  policies are deprecated and that a policy should be set to OLD only under
  specific short-term circumstances.  Projects should be ported to the NEW
  behavior and not rely on setting a policy to OLD.


-- Boost version: 1.59.0
-- Found the following Boost libraries:
--   date_time
--   filesystem
--   regex
--   signals
--   system
--   thread
--   log
--   chrono
--   atomic
--   log_setup
-- Could NOT find cppcheck (missing: CPPCHECK_EXECUTABLE) 
-- Could NOT find pycodestyle (missing: PYCODESTYLE_EXECUTABLE) 
-- Could NOT find Doxygen (missing: DOXYGEN_EXECUTABLE) 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/deepforest/Загрузки/src-tarball
```

**make**
```
[root@localhost src-tarball]# make -j2
[  0%] Building CXX object CMakeFiles/freeorionparseobj.dir/parse/CommonParamsParser.cpp.o
fatal: Not a git repository (or any of the parent directories): .git
WARNING: git not installed or not setup correctly
WARNING: Can't determine git commit!
Version matches version in existing Version.cpp, skip regenerating it
Building v0.4.8 build ???
[  0%] Built target freeorionversion
[  0%] Building CXX object GG/CMakeFiles/GiGi.dir/src/DropDownList.cpp.o
/home/deepforest/Загрузки/src-tarball/GG/src/DropDownList.cpp: В функции-члене «boost::optional<std::_List_iterator<std::shared_ptr<GG::ListBox::Row> > > ModalListPicker::KeyPressCommon(GG::Key, uint32_t, GG::Flags<GG::ModKey>)»:
/home/deepforest/Загрузки/src-tarball/GG/src/DropDownList.cpp:402:43: ошибка: нет подходящей функции для вызова «GG::ListBox::BringRowIntoView(std::initializer_list<std::_List_iterator<std::shared_ptr<GG::ListBox::Row> > >&)»
             LB()->BringRowIntoView(prev_it);
                                           ^
/home/deepforest/Загрузки/src-tarball/GG/src/DropDownList.cpp:402:43: замечание: candidate is:
In file included from /home/deepforest/Загрузки/src-tarball/GG/GG/DropDownList.h:34:0,
                 from /home/deepforest/Загрузки/src-tarball/GG/src/DropDownList.cpp:25:
/home/deepforest/Загрузки/src-tarball/GG/GG/ListBox.h:396:21: замечание: void GG::ListBox::BringRowIntoView(GG::ListBox::iterator)
     void            BringRowIntoView(iterator it);          ///< moves the scrollbars so that row \a it is visible
                     ^
/home/deepforest/Загрузки/src-tarball/GG/GG/ListBox.h:396:21: замечание:   no known conversion for argument 1 from «std::initializer_list<std::_List_iterator<std::shared_ptr<GG::ListBox::Row> > >» to «GG::ListBox::iterator {aka std::_List_iterator<std::shared_ptr<GG::ListBox::Row> >}»
/home/deepforest/Загрузки/src-tarball/GG/src/DropDownList.cpp:403:20: ошибка: could not convert «prev_it» from «std::initializer_list<std::_List_iterator<std::shared_ptr<GG::ListBox::Row> > >» to «boost::optional<std::_List_iterator<std::shared_ptr<GG::ListBox::Row> > >»
             return prev_it;
                    ^
make[2]: *** [GG/CMakeFiles/GiGi.dir/src/DropDownList.cpp.o] Ошибка 1
make[1]: *** [GG/CMakeFiles/GiGi.dir/all] Ошибка 2
make[1]: *** Ожидание завершения заданий...
[  1%] Building CXX object CMakeFiles/freeorionparseobj.dir/parse/ConditionParser1.cpp.o
[  1%] Building CXX object CMakeFiles/freeorionparseobj.dir/parse/ConditionParser7.cpp.o
[  1%] Building CXX object CMakeFiles/freeorionparseobj.dir/parse/ConditionParser.cpp.o
In file included from /usr/include/c++/4.8.2/memory:81:0,
                 from /usr/local/include/boost/config/no_tr1/memory.hpp:21,
                 from /usr/local/include/boost/get_pointer.hpp:14,
                 from /usr/local/include/boost/proto/transform/default.hpp:21,
                 from /usr/local/include/boost/phoenix/core/meta_grammar.hpp:17,
                 from /usr/local/include/boost/phoenix/core/is_nullary.hpp:15,
                 from /usr/local/include/boost/phoenix/core/actor.hpp:20,
                 from /usr/local/include/boost/phoenix/core.hpp:12,
                 from /usr/local/include/boost/spirit/include/phoenix_core.hpp:11,
                 from /usr/local/include/boost/spirit/home/support/terminal.hpp:16,
                 from /usr/local/include/boost/spirit/home/support/common_terminals.hpp:15,
                 from /usr/local/include/boost/spirit/home/lex/lexer/terminals.hpp:13,
                 from /usr/local/include/boost/spirit/home/lex/lexer.hpp:13,
                 from /usr/local/include/boost/spirit/home/lex.hpp:13,
                 from /usr/local/include/boost/spirit/home/lex/lexer_lexertl.hpp:15,
                 from /usr/local/include/boost/spirit/include/lex_lexertl.hpp:16,
                 from /home/deepforest/Загрузки/src-tarball/parse/Lexer.h:6,
                 from /home/deepforest/Загрузки/src-tarball/parse/ConditionParser.h:4,
                 from /home/deepforest/Загрузки/src-tarball/parse/ConditionParserImpl.h:4,
                 from /home/deepforest/Загрузки/src-tarball/parse/ConditionParser.cpp:1:
/usr/include/c++/4.8.2/bits/unique_ptr.h: In instantiation of «void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = ValueRef::ValueRefBase<int>]»:
/usr/include/c++/4.8.2/bits/unique_ptr.h:184:16:   required from «std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = ValueRef::ValueRefBase<int>; _Dp = std::default_delete<ValueRef::ValueRefBase<int> >]»
/usr/include/c++/4.8.2/bits/unique_ptr.h:157:61:   required from «constexpr std::unique_ptr<_Tp, _Dp>::unique_ptr(std::nullptr_t) [with _Tp = ValueRef::ValueRefBase<int>; _Dp = std::default_delete<ValueRef::ValueRefBase<int> >; std::nullptr_t = std::nullptr_t]»
/home/deepforest/Загрузки/src-tarball/parse/../universe/Condition.h:200:73:   required from here
/usr/include/c++/4.8.2/bits/unique_ptr.h:65:22: ошибка: invalid application of «sizeof» to incomplete type «ValueRef::ValueRefBase<int>»
  static_assert(sizeof(_Tp)>0,
                      ^
/usr/include/c++/4.8.2/bits/unique_ptr.h: In instantiation of «void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = ValueRef::ValueRefBase<double>]»:
/usr/include/c++/4.8.2/bits/unique_ptr.h:184:16:   required from «std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = ValueRef::ValueRefBase<double>; _Dp = std::default_delete<ValueRef::ValueRefBase<double> >]»
/usr/include/c++/4.8.2/bits/unique_ptr.h:157:61:   required from «constexpr std::unique_ptr<_Tp, _Dp>::unique_ptr(std::nullptr_t) [with _Tp = ValueRef::ValueRefBase<double>; _Dp = std::default_delete<ValueRef::ValueRefBase<double> >; std::nullptr_t = std::nullptr_t]»
/home/deepforest/Загрузки/src-tarball/parse/../universe/Condition.h:617:82:   required from here
/usr/include/c++/4.8.2/bits/unique_ptr.h:65:22: ошибка: invalid application of «sizeof» to incomplete type «ValueRef::ValueRefBase<double>»
/usr/include/c++/4.8.2/bits/unique_ptr.h: In instantiation of «void std::default_delete<_Tp>::operator()(_Tp*) const [with _Tp = ValueRef::ValueRefBase<std::basic_string<char> >]»:
/usr/include/c++/4.8.2/bits/unique_ptr.h:184:16:   required from «std::unique_ptr<_Tp, _Dp>::~unique_ptr() [with _Tp = ValueRef::ValueRefBase<std::basic_string<char> >; _Dp = std::default_delete<ValueRef::ValueRefBase<std::basic_string<char> > >]»
/usr/include/c++/4.8.2/bits/unique_ptr.h:157:61:   required from «constexpr std::unique_ptr<_Tp, _Dp>::unique_ptr(std::nullptr_t) [with _Tp = ValueRef::ValueRefBase<std::basic_string<char> >; _Dp = std::default_delete<ValueRef::ValueRefBase<std::basic_string<char> > >; std::nullptr_t = std::nullptr_t]»
/home/deepforest/Загрузки/src-tarball/parse/../universe/Condition.h:888:97:   required from here
/usr/include/c++/4.8.2/bits/unique_ptr.h:65:22: ошибка: invalid application of «sizeof» to incomplete type «ValueRef::ValueRefBase<std::basic_string<char> >»
make[2]: *** [CMakeFiles/freeorionparseobj.dir/parse/ConditionParser.cpp.o] Ошибка 1
make[2]: *** Ожидание завершения заданий...
make[1]: *** [CMakeFiles/freeorionparseobj.dir/all] Ошибка 2
make: *** [all] Ошибка 2
[root@localhost src-tarball]# 
```

**Comment by Dilvish-fo** [comment]
Current master builds just fine for me on Ubuntu Linux.  If you state somewhere in there what version of FO that tarball is, I'm not seeing it, please specify.  

Also, what version of GCC or Clang do you have?  [On Linux FO requires](https://github.com/freeorion/freeorion/blob/master/BUILD.md)
```
GNU GCC - 5.0 or later
Clang - 3.9 or later (4.0 or later on FreeBSD)
```
And, it looks to me like CENTOS 7 doesn't have that by default.
http://mirror.centos.org/centos/7/os/x86_64/Packages/

**Comment by ischwarze** [comment]
Here is more information on the failure.

Full build logs:
http://build-failures.rhaalovely.net//sparc64/2018-12-19/games/freeorion.log
http://build-failures.rhaalovely.net//powerpc/2018-12-14/games/freeorion.log

CPU architecture: occurs on both UltraSparc 64bit and on MacPPC
see https://www.openbsd.org/sparc64.html and https://www.openbsd.org/macppc.html

Operating system: OpenBSD 6.4-current
Compiler: GCC 4.9.4
Boost: 1.66.0
Release: FreeOrion_v0.4.8_2018-08-23.26f16b0_Source.tar.gz

Unfortunately, i no longer have a Sparc64 machine for testing (except for a test account on an Oracle Solaris 11 cluster - but i really don't feel like testing on Solaris, there are typically loads of additional portability problems to Solaris and FreeOrion does not appear to be all that robust in the first place) and i never had a MacPPC machine.  Anyway, i do not see this build failure on amd64 (= x86_64) ceteris paribus; same operating system, compiler, and boost.

I know that GCC 4.9.4 is lower than the GCC 5.0 you ask for - but it works with GCC 4.9.4 on amd64...  Why do you ask for at least 5.0?

Looking at the code in universe/Condition.h, it appears to use std::unique_ptr<ValueRef::ValueRefBase<int>>, but it only includes "ValueRefFwd.h" which only contains a forward declaration of "template <class T> struct ValueRefBase;", so the type appears to be incomplete, but std::unique_ptr<> appears to require a complete type because it appears to take the size internally.

Possibly, always including ValueRef.h before Condition.h might be required; but i'm not a C++ expert, so i may be mistaken.  If you understand the problem and are reasonably sure that you have a patch fixing it, i could probably pass that patch to other OpenBSD developers for testing.

**Comment by Dilvish-fo** [comment]
>  but std::unique_ptr<> appears to require a complete type because it appears to take the size internally

My understanding is that [std::unique_ptr<> only requires a complete type in a handful of cases](https://stackoverflow.com/a/6089065); I can't say that I've reviewed all our code for this particular situation though.

>  If you understand the problem and are reasonably sure that you have a patch fixing it,

I've got an understanding of what you are suggesting might be the problem, but so far it appears that this compiler error only appears on a very particular setup, one that doesn't even meet our min requirements, so it's not even clear to me that this is a valid bug.   I think that the dev who set that GCC 5.0 min req is not currently active, and I don't know their reasons, sorry; you might try checking for a discussion of it on our forums.

Since it seems that so far this error is only appearing on your non-req-compliant setup, and you have an idea of what you think might possibly clear it up, I'd suggest that you explore that yourself and let us know if it does indeed resolve it.
